rest-service/manager_rest/deployment_update/step_extractor.py

```diff
@@ -0,0 +1,306 @@
+            if node_name in old_nodes:
+                old_node = old_nodes[node_name]
+
+                if node[TYPE] != old_node[TYPE]:
+                    # a node that changed its type counts as a different node
+                    self._create_deployment_update_step(NODE)
+                    # Since the node changed its type, there is no need to
+                    # compare its other fields.
+                    del self.entity_id[-1]
+                    continue
+
+                self._extract_steps_from_operations(
+                    operations_type_name=OPERATIONS,
+                    operations=node[OPERATIONS],
+                    old_operations=old_node[OPERATIONS])
+
+                self._extract_steps_from_relationships(
+                    relationships=node[RELATIONSHIPS],
+                    old_relationships=old_node[RELATIONSHIPS])
+
+                self._extract_steps_from_entities(
+                    PROPERTIES,
+                    new_entities=node[PROPERTIES],
+                    old_entities=old_node[PROPERTIES])
+
+            else:
+                # add (remove) node step
+                self._create_deployment_update_step(NODE)
+            del self.entity_id[-1]
+        del self.entity_id[-1]
+
+    def _extract_steps_from_entities(self,
+                                     entities_id,
+                                     new_entities,
+                                     old_entities):
+        # Dan suggests to rename `entities_id` to something that shows its
+        # role as a part of the EntityID, and to add another parameter,
+        # `entity_type`, that will eliminate the need for `_to_singular_from`.
+
+        self.entity_id.append(entities_id)
+        for entity_name in new_entities:
+            self.entity_id.append(entity_name)
+            if entity_name in old_entities:
+                if old_entities[entity_name] != new_entities[entity_name]:
+                    self._create_deployment_update_step(
+                        _to_singular_form(entities_id), modify=True)
+            else:
+                self._create_deployment_update_step(
+                    _to_singular_form(entities_id))
+
+            del self.entity_id[-1]
+        del self.entity_id[-1]
+
+    def _extract_steps(self, new, old):
+
+        for entities_name, new_entities in new.iteritems():
+            old_entities = old.get(entities_name, {})
+
+            if entities_name == NODES:
+                self._extract_steps_from_nodes(
+                    new_entities, old_entities)
+
+            elif entities_name == OUTPUTS:
+                self._extract_steps_from_entities(
+                    OUTPUTS, new_entities, old_entities)
+
+            elif entities_name == WORKFLOWS:
+                self._extract_steps_from_entities(
+                    WORKFLOWS, new_entities, old_entities)
+
+    def _create_deployment_update_step(self, entity_type, modify=False):
```

Comment:
A wrapper function that captures the logic of determining the action. If there is an action, just call _create_deployment_update_step (that just calls create_deployment_update_step)